rclcpp_components/doc/architecture.rst

```diff
@@ -0,0 +1,263 @@
+However, you can avoid this macro and the ``rclcpp_components_describe_node_component()`` macro calls by creating your own plugin description XML file and passing it with the ``PLUGIN_DESCRIPTION_FILE <file path>`` option.
+
+So, the required steps are registering your nodes in the source code, creating a shared library, and registering with ``pluginlib``.
+However, you have some choice in how the ``pluginlib`` description file is created and how registration occurs, plus creating the node executables is completely optional.
+
+There are more options available to you, and if you do not wish to use the convenience macros in CMake you can see in the details that follow how each are implemented and do any of the steps manually yourself.
+
+Writing the Code for a Node Component
+-------------------------------------
+
+Consider this trivial example:
+
+.. code-block:: cpp
+
+    #include <rclcpp/rclcpp.hpp>
+
+    class MyNode : public rclcpp::Node
+    {
+    public:
+      MyNode(rclcpp::NodeOptions options) : rclcpp::Node("my_node", options) {}
+    };
+
+    #include <rclcpp_components/register_node_macro.hpp>
+
+    RCLCPP_COMPONENTS_REGISTER_NODE(MyNode)
+
+The contract for the class passed to ``RCLCPP_COMPONENTS_REGISTER_NODE`` is that it:
+
+- has a constructor which takes a single argument which is an instance of ``rclcpp::NodeOptions``
+- has a method of the signature ``rclcpp::node_interfaces::NodeBaseInterface::SharedPtr get_node_base_interface()``
+
+Note that it does not *need* to inherit from ``rclcpp::Node``, but that is the easiest way to do it and to get the ``get_node_base_interface()`` method for "free".
+
+The RCLCPP_COMPONENTS_REGISTER_NODE Macro
++++++++++++++++++++++++++++++++++++++++++
+
+One possible (psuedo-code) implementation for the ``RCLCPP_COMPONENTS_REGISTER_NODE`` macro might be:
+
+.. code-block:: cpp
+
+    #define RCLCPP_COMPONENTS_REGISTER_NODE(NodeClass) \
+      CLASS_LOADER_REGISTER_CLASS(NodeFactoryTemplate<NodeClass>, rclcpp_components::NodeFactory)
+
+Where the base class ``rclcpp_components::NodeFactory`` is the actual plugin interface and the ``rclcpp_components::NodeFactoryTemplate`` class is a templated "adapter" class which will provide the ``rclcpp_components::NodeFactory`` interface given a sub-class of ``rclcpp::Node``.
+
+The NodeFactory and Related Interfaces
+++++++++++++++++++++++++++++++++++++++
+
+The ``rclcpp_components::NodeFactory`` interface is dual-purposed, first it works around a limitation of instantiating classes via a general purpose class loading interface which is that you cannot call constructors on derived classes because you only know the base class and are unaware of any different signatures for constructors in the derived class.
```

Comment:
```suggestion
The ``rclcpp_components::NodeFactory`` interface is dual-purposed.
First, it works around a limitation of instantiating classes via a general purpose class loading interface which is that you cannot call constructors on derived classes because you only know the base class and are unaware of any different signatures for constructors in the derived class.
```